datalad/local/subdatasets.py

```diff
@@ -284,21 +281,33 @@ def _get_submodules(ds, paths, fulfilled, recursive, recursion_limit,
     repo = ds.repo
     if not GitRepo.is_valid_repo(dspath):
         return
+    # expand all test cases for the contains test in the loop below
+    # leads to ~20% speedup per loop iteration of a non-match
+    expanded_contains = [
+        # convert to strings, as sm_path below will be a string and
+        # we do not want repeated type-conversion
+        [c] + list(c.parents)
+        for c in (contains if contains is not None else [])
+    ]
     # put in giant for-loop to be able to yield results before completion
     for sm in _parse_git_submodules(ds.pathobj, repo, paths):
-        contains_hits = []
+        sm_path = sm['path']
+        contains_hits = None
         if contains:
-            contains_hits = [
-                c for c in contains if sm['path'] == c or sm['path'] in c.parents
-            ]
+            contains_hits = [c[0] for c in expanded_contains if sm_path in c]
             if not contains_hits:
                 # we are not looking for this subds, because it doesn't
                 # match the target path
                 continue
+        # the following used to be done by _parse_git_submodules()
+        # but is expensive and does not need to be done for submodules
+        # no matching `contains`
```

Comment:
s/no/not/

Reply:
Pushed a fixup for the failing tests, so I took care of this as well.

Reply:
Thx!